File: solutions/ocaml/059-decode-ways.ml
(*
  Problem 59: Decode Ways (LeetCode 91)
  Difficulty: Med
  Language: OCaml
*)
let num_decodings s =
  let n = String.length s in
  let rec go i memo =
    if i = n then (1, memo)
    else if s.[i] = '0' then (0, memo)
    else match List.assoc_opt i memo with
      | Some v -> (v, memo)
      | None ->
        let (a, m1) = go (i+1) memo in
        let (b, m2) =
          if i+1 < n then
            let v = (Char.code s.[i] - 48) * 10 + (Char.code s.[i+1] - 48) in
            if v <= 26 then go (i+2) m1 else (0, m1)
          else (0, m1)
        in (a+b, (i, a+b) :: m2)
  in fst (go 0 [])

File: solutions/haskell/059-decode-ways.hs
{-
  Problem 59: Decode Ways (LeetCode 91)
  Difficulty: Med
  Language: Haskell
-}
import Data.Map.Strict (Map, empty, lookup, insert)

numDecodings :: String -> Int
numDecodings s = fst (go 0 empty)
  where
    n = length s
    go i memo
      | i == n = (1, memo)
      | s !! i == '0' = (0, memo)
      | otherwise = case Data.Map.Strict.lookup i memo of
          Just v -> (v, memo)
          Nothing ->
            let (a, m1) = go (i+1) memo
                (b, m2) = if i+1 < n && read (take 2 (drop i s)) <= (26::Int)
                           then go (i+2) m1 else (0, m1)
                res = a + b
            in (res, insert i res m2)
